### PR TITLE
compare TypeToken with fullDispString between CompositeTypeToken and JvmTypeToken

### PR DESCRIPTION
**Pull request by salongbass**
Please see Issue about this

### CODE REVIEW

**Comment by SalomonBrys** [comment]
Can you explain this MR ?

File: shared/core/kotlin/com/github/salomonbrys/kodein/typeToken.kt [hunk 1]
```diff
@@ -83,9 +84,9 @@ class CompositeTypeToken<T>(val main: TypeToken<T>, vararg val params: TypeToken
             throw IllegalStateException("CompositeTypeToken must be given at least one type parameter")
     }
 
-    override fun simpleDispString() = "${main.simpleDispString()}<${params.joinToString(", ")}>"
+    override fun simpleDispString() = "${main.simpleDispString()}<${params.joinToString(", ") { it.simpleDispString() }}>"
 
-    override fun fullDispString() = "${main.fullDispString()}<${params.joinToString(", ")}>"
+    override fun fullDispString() = "${main.fullDispString()}<${params.joinToString(", ") { it.fullDispString() }}>"
```
Comment (salongbass) [inline shared/core/kotlin/com/github/salomonbrys/kodein/typeToken.kt]: This is a bug for show *DispString

File: kodein-core/src/main/kotlin/com/github/salomonbrys/kodein/types.kt [hunk 2]
```diff
@@ -69,16 +69,21 @@ internal abstract class JVMTypeToken<T> : TypeToken<T> {
 
     override fun equals(other: Any?): Boolean {
         if (this === other) return true
+        if (other is CompositeTypeToken<*>) {
+            return fullDispString().equals(this.fullDispString())
```
Comment (salongbass) [inline kodein-core/src/main/kotlin/com/github/salomonbrys/kodein/types.kt]: compare between general type token and composite type tonken with their's fullDispString.
generic<A<B>>() == erasedComp1<A<B>>(erased<A>(), erased<B>())

File: kodein-core/src/main/kotlin/com/github/salomonbrys/kodein/types.kt [hunk 3]
```diff
@@ -69,16 +69,21 @@ internal abstract class JVMTypeToken<T> : TypeToken<T> {
...
 
         if (type() != other.type()) return false
 
         return true
     }
 
-    override fun hashCode(): Int {
-        return type().hashCode()
-    }
+    override fun hashCode() = fullDispString().hashCode()
```
Comment (salongbass) [inline kodein-core/src/main/kotlin/com/github/salomonbrys/kodein/types.kt]: for CMap hash find use fullDispString's hashcode.
So, generic&lt;A&lt;B&gt;&gt;().hashCode() == erasedComp1&lt;A&lt;B&gt;&gt;(erased(), erased&lt;B&gt;()).hashCode()

File: shared/core/kotlin/com/github/salomonbrys/kodein/typeToken.kt (hunk 1 above)
Comment (SalomonBrys) [reply shared/core/kotlin/com/github/salomonbrys/kodein/typeToken.kt]: This is out of date.

File: kodein-core/src/main/kotlin/com/github/salomonbrys/kodein/types.kt (hunk 2 above)
Comment (SalomonBrys) [reply kodein-core/src/main/kotlin/com/github/salomonbrys/kodein/types.kt]: This is very interesting !

File: kodein-core/src/main/kotlin/com/github/salomonbrys/kodein/types.kt (hunk 3 above)
Comment (SalomonBrys) [reply kodein-core/src/main/kotlin/com/github/salomonbrys/kodein/types.kt]: This is very interesting !

**Comment by SalomonBrys** [comment]
So, this PR is very interresting.
However, is it still applicable to the "new" type tokens ?

**Pull request closed without merging**